formbricks/formbricks
# feat: SingleUseSurveyLinks

## What does this PR do?

This PR adds a single-use link option to Link Surveys. 

Added a suId field to the Response prisma model that is unique per survey.
Added a singleUse object in the survey prisma model to stored information about the single use options.

Added an extra option for enabling single-use links in the Response Options section when creating and editing surveys, here users can also modify the heading and subheading in the page that shows the user that the link has already been used.

In the survey summary, added a single use id to each of the response.

When single use survey is enabled, The link survey modal is modified to generate a link with an suId query parameter so users and send a working link right away, replace the embed button with a generate link button to generate new valid links.

Created a SurveyLinkUsed component to display when users open a link with a existing response for a single use survey.

Added additional logic in the server component for link surveys to check if the link survey has the single use option enabled, and to check if the SurveyLinkUsed should be displayed or not.


Fixes #72 
https://www.loom.com/share/f0a4a98a4f3f416ea74b0e076463adb6?sid=3f23ebc2-b22b-401d-9c3b-151dfc303d43

## Type of change


## How should this be tested?

1. Create a survey with alteast two questions (for data pre-filling).
2. enable the single-use link option in the survey settings.
3. Use the survey link with suId query parameter to create responses.
4. Once the survey is completed with a link, try to reuse that link and expect the app to show a page stating the the link has been used already.
5. Generate new single-use links in from the summary page, by pressing the generate button in link survey.
6. Test survey link with userId and data-prefilling query parameters (each individually and together).
7. Try removing the singe-use option in the survey settings.
8. The survey should work multiple times with any link

## Checklist
